### PR TITLE
Use a separate `TextPainter` for intrinsics calculation in `RenderEditable` and `RenderParagraph`

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Use a dedicated `TextPainter` for intrinsic size calculation in `RenderEditable` and `RenderParagraph`.

This is an implementation detail so the change should be covered by existing tests.  Performance wise this shouldn't be significantly slower since SkParagraph [caches the result of slower operations across different paragraphs](https://github.com/google/skia/blob/9c62e7b382cf387195ef82895530c97ccceda690/modules/skparagraph/src/ParagraphCache.cpp#L254-L272). Existing benchmarks should be able to catch potential regressions (??).

The reason for making this change is to make sure that intrinsic size computations don't destroy text layout artifacts, so I can expose the text layout as a stream of immutable `TextLayout` objects, to signify other render objects that text-layout-dependent-cache (such as caches for `getBoxesForRange` which can be relatively slow to compute) should be invalidated and  `markNeedsPaint` needs to be called if the painting logic depended on text layout.
Without this change, the intrinsics/dry layout calculations will add additional events to the text layout stream, which violates the "dry"/non-destructive contract.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by LongCatIsLooong** [comment]
@goderbauer I'm not 100% sure this is covered by existing benchmarks. If not what kind of benchmarks do you think would be helpful? One with `Text` in an `IntrinsicWidth/Height` and keeps resizing?

**Comment by goderbauer** [comment]
> I'm not 100% sure this is covered by existing benchmarks.

As far as I know we don't really have a benchmark covering these cases.

> One with Text in an IntrinsicWidth/Height and keeps resizing?

That seems like a good idea to cover this change.

**Comment by LongCatIsLooong** [comment]
I ran the newly added microbenchmark locally and didn't observe a performance regression. The Google testing failures are scuba changes

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -803,18 +778,33 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
...
+      ..textAlign = _textPainter.textAlign
+      ..textDirection = _textPainter.textDirection
+      ..textScaler = _textPainter.textScaler
+      ..maxLines = _textPainter.maxLines
+      ..ellipsis = _textPainter.ellipsis
+      ..locale = _textPainter.locale
+      ..strutStyle = _textPainter.strutStyle
+      ..textWidthBasis = _textPainter.textWidthBasis
+      ..textHeightBehavior = _textPainter.textHeightBehavior;
+  }
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart:798]: I wonder if we could somehow have a test to verify that the two text painters remain in sync? I don't have any ideas of how to do that, though.

File: packages/flutter/test/rendering/paragraph_test.dart [hunk 2]
```diff
@@ -375,6 +375,27 @@ void main() {
...
+    expect(getRectForA(), const Rect.fromLTWH(0, 0, 10, 10));
+
+    paragraph.textAlign = TextAlign.right;
+    expect(paragraph.debugNeedsLayout, isFalse);
+    expect(paragraph.debugNeedsPaint, isTrue);
+
+    paragraph.paint(MockPaintingContext(), Offset.zero);
+    expect(getRectForA(), const Rect.fromLTWH(0, 0, 10, 10));
+  });
+
```
Comment (goderbauer) [inline packages/flutter/test/rendering/paragraph_test.dart:398]: Do we have enough existing tests to cover that when you change the configuration of a RenderParagraph or RenderEditable that the new configuration is properly accounted for in the intrinsics calculations?

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by goderbauer** [comment]
> The Google testing failures are scuba changes

Wait, why does this change affect any goldens?

File: packages/flutter/test/rendering/paragraph_test.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/rendering/paragraph_test.dart:398]: Added a few tests.  I couldn't think of ways to test some configurations using the test font (`textDirection` for example).

**Comment by LongCatIsLooong** [comment]
> > The Google testing failures are scuba changes
> 
> Wait, why does this change affect any goldens? Isn't that kinda unexpected?

`TextPainter` sometimes tries to lay out the text using the text's maxIntrinsic width instead of the real input width, and I think that could cause the pixel shifts in the scubas.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:798]: I was hoping the dry size check in `RenderBox.size` setter would help catch inconsistencies in the two text painters?

**Comment by TahaTesser** [comment]
@LongCatIsLooong 
My PR https://github.com/flutter/flutter/pull/143932 fails web HTML renderer tests after this PR and potentially also failing https://github.com/flutter/flutter/pull/142804 tests.

**Comment by LongCatIsLooong** [comment]
> @LongCatIsLooong My PR #143932 fails web HTML renderer tests after this PR and potentially also failing #142804 tests.

I'll take a look in a bit.

**Pull request merged**